tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala

```diff
@@ -66,6 +69,20 @@ final class ScalaJSClassEmitter private (
     linkingUnit.classDefs.map(c => c.encodedName -> c).toMap
   }
 
+  private[javascript] lazy val hasSubclasses: Set[String] = {
+    if (linkingUnit == null) {
+      throw new IllegalArgumentException(
+          "A class emitter created without a LinkingUnit cannot emit JS classes")
+    }
+    linkingUnit.classDefs.map(c => c.superClass).
+      collect{case Some(x) => x.name}.toSet
+  }
+
+  def effectivelyFinal(clss: LinkedClass): Boolean = {
```

Comment:
This method could actually made a `lazy val Set` instead of `hasSubclasses`. `hasSubclasses` is only used in this method.

Moreover, calls to this method are always combined with a test for the number of constructors. I suggest incorporating this information in the result, and therefore rename it as something like `usesJSConstructorOpt`, clearly stating what this def/set is for: the set of classes for which we use the JS constructor optimization.